presto-spark-base/src/main/java/com/facebook/presto/spark/planner/PrestoSparkRddFactory.java

```diff
@@ -235,38 +237,25 @@ private static Partitioner createPartitioner(PartitioningHandle partitioning, in
                             taskStatsCollector,
                             toTaskProcessorBroadcastInputs(broadcastInputs)));
         }
-        else if (rddInputs.size() == 1) {
-            Entry<PlanFragmentId, JavaPairRDD<Integer, PrestoSparkRow>> input = getOnlyElement(rddInputs.entrySet());
-            PairFlatMapFunction<Iterator<Tuple2<Integer, PrestoSparkRow>>, Integer, PrestoSparkRow> taskProcessor =
-                    createTaskProcessor(
-                            executorFactoryProvider,
-                            serializedTaskDescriptor,
-                            input.getKey().toString(),
-                            taskStatsCollector,
-                            toTaskProcessorBroadcastInputs(broadcastInputs));
-            return input.getValue()
-                    .mapPartitionsToPair(taskProcessor);
-        }
-        if (rddInputs.size() == 2) {
-            List<PlanFragmentId> fragmentIds = ImmutableList.copyOf(rddInputs.keySet());
-            List<JavaPairRDD<Integer, PrestoSparkRow>> rdds = fragmentIds.stream()
-                    .map(rddInputs::get)
-                    .collect(toImmutableList());
-            FlatMapFunction2<Iterator<Tuple2<Integer, PrestoSparkRow>>, Iterator<Tuple2<Integer, PrestoSparkRow>>, Tuple2<Integer, PrestoSparkRow>> taskProcessor =
-                    createTaskProcessor(
-                            executorFactoryProvider,
-                            serializedTaskDescriptor,
-                            fragmentIds.get(0).toString(),
-                            fragmentIds.get(1).toString(),
-                            taskStatsCollector,
-                            toTaskProcessorBroadcastInputs(broadcastInputs));
-            return JavaPairRDD.fromJavaRDD(
-                    rdds.get(0).zipPartitions(
-                            rdds.get(1),
-                            taskProcessor));
+
+        ImmutableList.Builder<String> fragmentIds = ImmutableList.builder();
+        ImmutableList.Builder<RDD<Tuple2<Integer, PrestoSparkRow>>> rdds = ImmutableList.builder();
+        for (Entry<PlanFragmentId, JavaPairRDD<Integer, PrestoSparkRow>> input : rddInputs.entrySet()) {
+            fragmentIds.add(input.getKey().toString());
+            rdds.add(input.getValue().rdd());
         }
 
-        throw new IllegalArgumentException(format("unsupported number of inputs: %s", rddInputs.size()));
+        Function<List<Iterator<Tuple2<Integer, PrestoSparkRow>>>, Iterator<Tuple2<Integer, PrestoSparkRow>>> taskProcessor = createTaskProcessor(
+                executorFactoryProvider,
+                serializedTaskDescriptor,
+                fragmentIds.build(),
+                taskStatsCollector,
+                toTaskProcessorBroadcastInputs(broadcastInputs));
+
+        return JavaPairRDD.fromRDD(
+                new PrestoSparkZipRdd(sparkContext.sc(), rdds.build(), taskProcessor),
+                classTag(Integer.class),
```

Comment:
My understanding Spark Java API avoids directly using `scala.reflect.ClassTag$.MODULE$`,  and that's why `JavaRDDLike` is introduced with automatic wrap of class tags, e.g. https://github.com/apache/spark/blob/fd2d55c9919ece5463377bc6f45f2cdb8bf90515/core/src/main/scala/org/apache/spark/api/java/JavaRDDLike.scala#L93

In fact, in `JavaRDDLike`, it use "`fakeClassTag`"  to "keep ClassTags out of the external Java API", as explained here: https://github.com/apache/spark/blob/fd2d55c9919ece5463377bc6f45f2cdb8bf90515/core/src/main/scala/org/apache/spark/api/java/JavaSparkContext.scala#L754-L764

```
   * Produces a ClassTag[T], which is actually just a casted ClassTag[AnyRef].
   *
   * This method is used to keep ClassTags out of the external Java API, as the Java compiler
   * cannot produce them automatically. While this ClassTag-faking does please the compiler,
   * it can cause problems at runtime if the Scala API relies on ClassTags for correctness.
   *
   * Often, though, a ClassTag[AnyRef] will not lead to incorrect behavior, just worse performance
   * or security issues. For instance, an Array[AnyRef] can hold any type T, but may lose primitive
   * specialization.
```

cc @sameeragarwal

Reply:
`$.MODULE$` Is the way Scala implements lazy singleton objects. Although this is not very "public" API, given the legacy I don't expect it to change.

Reply:
@arhimondr : The fundamental question is how to interact between Java/Scala API in such case. I assume Databricks has done in-depth investigation into this when they design the Spark Java API. Thus there might be reason why they use `fakeClassTag` instead of `scala.reflect`.

I will check this offline with @sameeragarwal . In the meanwhile, we can continue with what is implemented in your PR. Just keep in mind we might need more investigation in the future about Java-Scala interaction (which I heard can be tricky in some cases due to Scala has many fancy features done by Scala Compiler 😃 ) and pay back tech-debt, if there is 😃